File: app/src/vdsensor/esp3/response.py
"""RESPONSE (packet type 0x02) parser (§2.2)."""

from __future__ import annotations

from dataclasses import dataclass

from .framing import Frame
from .packets import PacketType, ReturnCode


@dataclass(frozen=True)
class Response:
    return_code: int
    payload: bytes                  # any extra response bytes after the return code
    opt: bytes

    @property
    def ok(self) -> bool:
        return self.return_code == ReturnCode.RET_OK


def parse_response(frame: Frame) -> Response:
    if frame.packet_type != PacketType.RESPONSE:
        raise ValueError(f"not a RESPONSE frame: type={frame.packet_type:#x}")
    if not frame.data:
        raise ValueError("RESPONSE frame has empty data")
    return Response(return_code=frame.data[0], payload=frame.data[1:], opt=frame.opt)

File: app-rs/src/esp3/response.rs
//! RESPONSE (packet type 0x02) parser (§2.2).

use super::framing::Frame;
use super::{ParseError, PacketType, ReturnCode};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Response {
    pub return_code: u8,
    /// Extra response bytes after the return code.
    pub payload: Vec<u8>,
    pub opt: Vec<u8>,
}

impl Response {
    pub fn ok(&self) -> bool {
        self.return_code == ReturnCode::RetOk as u8
    }
}

/// Parse a RESPONSE frame.
///
pub fn parse_response(frame: &Frame) -> Result<Response, ParseError> {
    if frame.packet_type != PacketType::Response as u8 {
        return Err(ParseError::WrongPacketType {
            expected: PacketType::Response as u8,
            actual: frame.packet_type,
        });
    }
    if frame.data.is_empty() {
        return Err(ParseError::EmptyData);
    }
    Ok(Response {
        return_code: frame.data[0],
        payload: frame.data[1..].to_vec(),
        opt: frame.opt.clone(),
    })
}
